### ISSUE TITLE
Modification doesn't get synced/stored

### ISSUE THREAD

**Issue by pmeisen** [orientechnologies/orientdb#6541]
## OrientDB Version
- [x] v2.2
## Operating System
- [x] MacOSX
## Expected behavior and actual behavior

I need some advice on how to figure out the following problem. I have an empty db (noTx) using the following configuration:

```
factory.setUseLightweightEdges(true);
factory.setUseClassForEdgeLabel(false);
factory.setUseClassForVertexLabel(false);
factory.setKeepInMemoryReferences(false);
factory.setAutoStartTx(false);

graph = factory.getNoTx();
```

I do some modification on the database, add, lookup, update, and removed vertices and edges. After a while I print out the current state with:

```
graph.getVertices().forEach(v -> {
    final ORecord rec = graph.getRawGraph().getRecord(OIdentifiable.class.cast(v.getId()));
    System.out.println("----> " + rec.getIdentity());
    v.getEdges(Direction.IN).forEach(e -> {
        System.out.println("   - IN: " + e);
    });
    v.getEdges(Direction.OUT).forEach(e -> {
        System.out.println("   - OUT: " + e);
    });
});
```

and receive:

```
----> #65:0
   - IN: e[#105:0->#65:0][#105:0-EDGEA->#65:0]
----> #66:0
   - IN: e[#145:0->#66:0][#145:0-EDGEB->#66:0]
----> #67:0
   - IN: e[#145:0->#67:0][#145:0-EDGEA->#67:0]
----> #105:0
   - OUT: e[#105:0->#145:0][#105:0-EDGEC->#145:0]
   - OUT: e[#105:0->#65:0][#105:0-EDGEA->#65:0]
----> #145:0
   - IN: e[#105:0->#145:0][#105:0-EDGEC->#145:0]
   - OUT: e[#145:0->#67:0][#145:0-EDGEA->#67:0]
   - OUT: e[#145:0->#66:0][#145:0-EDGEB->#66:0]
```

everything is ok like that, but if I look at the running database using the UI or if I start the output aside in a different thread the result is:

```
----> #65:0
   - IN: e[#105:0->#65:0][#105:0-EDGEA->#65:0]
----> #66:0
   - IN: e[#145:0->#66:0][#145:0-EDGEB->#66:0]
----> #67:0
   - IN: e[#145:0->#67:0][#145:0-EDGEA->#67:0]
----> #105:0
   - OUT: e[#105:0->#145:0][#105:0-EDGEC->#145:0]
   - OUT: e[#105:0->#65:0][#105:0-EDGEA->#65:0]
----> #145:0
   - IN: e[#105:0->#145:0][#105:0-EDGEC->#145:0]
   - OUT: e[#145:0->#66:0][#145:0-EDGEB->#66:0]
```

You may notice, that the last edge (#145:0 is missing an OUT edge, the IN edge is there #67:0). This addEdge is the last statement I fire utilizing the Java API, if I add more edges or remove some the operations (I only tested a couple) are "synced" across the UIs and threads, just the one addEdge is somehow broken. 

```
----> #145:0
   - OUT: e[#145:0->#67:0][#145:0-EDGEA->#67:0]
```

Can anybody point me in any direction what the problem could be (storage?, do I have to commit sometimes, even using noTx, ....)? How can I debug the problem (I'm using Java)? How can I see what the underlying problem may be? Why is the edge not being synced? Where does the sync with whatever happen? I'm just utilizing the Java API nothing special.

Thanks!

PS: I'm using a locally running OrientDB server (creating a plocal database)
PPS: All the operations I do are single-threaded, so no multi-threading involved, the only operation I run in a different thread (with a different database instance) is the sout, which runs only once after all other operations are applied.

**Comment by pmeisen** [comment]
I did some further investigation and I'm pretty sure that may be a bug (or a possible miss-usage which should throw an exception). The record behind the #145:0 is not getting saved, it just gets modified (the version does not change). The reason for that is, that the record is not marked as updated by the `OTransactionNoTx::saveRecord` implementation:

```
ODirtyManager dirtyManager = ORecordInternal.getDirtyManager(iRecord);
Set<ORecord> newRecord = dirtyManager.getNewRecords();
Set<ORecord> updatedRecord = dirtyManager.getUpdateRecords();
```

I couldn't figure out why the `dirtyManager` returns `null` in that specific case. I'll investigate further and keep the post updated!

**Comment by pmeisen** [comment]
Ok further things I found. It looks like that the vertex is not marked as updated, because the following happens in `OrientVertex`:

1. The following lines are executed for the specified vertex (#145:0), which sets the `out = null`
```
} else if (found instanceof ORidBag) {
  // ADD THE LINK TO THE COLLECTION
  out = null;
  ((ORidBag) found).add(iTo);
} 
```

2. The overwrite is not executed, which normally ensures that the `setDirty()` is executed
```
if (out != null)
  // OVERWRITE IT
  iFromVertex.field(iFieldName, out, outType);
```

3. The vertex is modified in the "cache", but not written to the database.

So it looks like it is a bug... any information?

**Comment by pmeisen** [comment]
Last finding, if I manipulate a property of vertex #145:0, the dirty flag gets set, and the data (including the edge information) is correctly synced. So now I would call this a bug for sure, please confirm.

```
final OrientEdge edge = graph.addEdge(null, sourceVertex, targetVertex, label);
sourceVertex.setProperty("RND", new Random().nextInt());
```

**Comment by tglman** [comment]
hi @pmeisen,

Nice detailed investigation, yes if you change a property on the record would make it dirty and persist it, in any case also when `((ORidBag) found).add(iTo);` happen there should be a callback inside that put the record in a status of "dirty" or  "dirty not changed" and make it finish inside the "updatedRecord" Set, and we have a lot of tests around this, but i think you may found a conbination of flags that create some problem, do you have a sequence of operation that can reproduce the problem ?

Thank you

**Comment by tglman** [comment]
hi @pmeisen,

We could reproduce this problem with another user that provided the test case, we fixed it and will be released in 2.2.13.

Bye

**Comment by tglman** [comment]
fixed and released, closing
